File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        res=[]
        top=0
        l=0
        r=len(matrix[0])-1
        b=len(matrix)-1
        while top<=b and l<=r:
            for i in range(l,r+1):
                res.append(matrix[top][i])
            top+=1
            for i in range(top,b+1):
                res.append(matrix[i][r])
            r-=1
            if top<=b:
                for i in range(r,l-1,-1):
                    res.append(matrix[b][i])
                b-=1
            if l<=r:
                for i in range(b,top-1,-1):
                    res.append(matrix[i][l])
                l+=1
        return res

File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
       int m=matrix.length;
       int n=matrix[0].length; 
       int srow=0,scol=0,erow=m-1,ecol=n-1;
       List<Integer> res=new ArrayList<>();
       while(srow<=erow && scol<=ecol)
       {
            for(int i=scol;i<=ecol;i++)
            {
                res.add(matrix[srow][i]);
            }
            for(int i=srow+1;i<=erow;i++)
            {
                res.add(matrix[i][ecol]);
            }
            for(int i=ecol-1;i>=scol;i--)
            {
                if(srow==erow)
                {
                    break;
                }
                res.add(matrix[erow][i]);
            }
            for(int i=erow-1;i>=srow+1;i--)
            {
                if(scol==ecol)
                {
                    break;
                }
                res.add(matrix[i][scol]);
            }
            srow++;scol++;erow--;ecol--;
       }
       return res;
    }
}